src/server_name_resolver.rs

```diff
@@ -0,0 +1,44 @@
+use std::error::Error;
+
+use http::Uri;
+use pki_types::ServerName;
+
+/// A trait implemented by types that can resolve a [`ServerName`] for a request.
```

Comment:
Since the `connector` module is pretty short anyway, I'd prefer to stick all of this stuff in the bottom of that module. Note that we prefer a top-down style, so please order as such:

```rust
1. `struct FixedServerNameResolver` + impls
2. `impl<F, E> ResolveServerName for F`
3. `pub trait ResolveServerName`
```